journal.py

```diff
@@ -91,43 +93,49 @@ def __repr__(self):
 # all the views
 
 
-@view_config(route_name='add', request_method='POST')
-def add_entry(request):
-    title = request.params.get('title')
-    text = request.params.get('text')
-    Entry.write(title=title, text=text)
-    return HTTPFound(request.route_url('home'))
-
-
-@view_config(route_name='add', renderer="templates/add.jinja2")
+@view_config(route_name='add', xhr=True, renderer='json')
+@view_config(route_name='add', xhr=False, renderer="templates/add.jinja2")
 def add_entry_view(request):
-    username = request.params.get('username', '')
-    return {'username': username}
+    if not request.authenticated_userid:
+        return HTTPFound(request.route_url('login'))
```

Comment:
I don't think you want to do this redirect if the request came via JSON.  It will not work the way you expect it to.  The ajax query in javascript will receive a redirect response, but it will not follow it and it won't raise an error as you might expect, either.  It would be better to raise `HTTPForbidden` if the request was ajax-borne